ksql-api/src/main/java/io/confluent/ksql/api/server/BufferedPublisher.java

```diff
@@ -0,0 +1,284 @@
+    }
+    return buffer.size() >= bufferMaxSize;
+  }
+
+  /**
+   * If you set a drain handler. It will be called if, after delivery is attempted there are zero
+   * elements buffered internally and there is demand from the subscriber for more elements.
+   *
+   * @param handler The handler
+   */
+  public void drainHandler(final Runnable handler) {
+    checkContext();
+    this.drainHandler = handler;
+  }
+
+  /**
+   * Mark the incoming stream of elements as complete. This means onComplete will be called on any
+   * subscriber after any buffered messages have been delivered. Once complete has been called no
+   * further elements will be accepted
+   */
+  public void complete() {
+    checkContext();
+    if (cancelled || complete) {
+      return;
+    }
+    completing = true;
+    if (buffer.isEmpty() && subscriber != null) {
+      sendComplete();
+    } else {
+      complete = true;
+    }
+  }
+
+  /**
+   * Subscribe a subscriber to this publisher. The publisher will allow at most one subscriber.
+   *
+   * @param subscriber The subscriber
+   */
+  @Override
+  public void subscribe(final Subscriber<? super T> subscriber) {
+    Objects.requireNonNull(subscriber);
+    if (Vertx.currentContext() == ctx) {
+      doSubscribe(subscriber);
+    } else {
+      ctx.runOnContext(v -> doSubscribe(subscriber));
+    }
+  }
+
+  /**
+   * Hook to allow subclasses to inject errors etc.
+   * This will be called before onNext is called on the subscriber to deliver an element
+   *
+   * @return true if processing should continue
+   */
+  protected boolean beforeOnNext() {
+    return true;
+  }
+
+  protected void sendError(final Exception e) {
+    checkContext();
+    try {
+      subscriber.onError(e);
+      cancelled = true;
+    } catch (Throwable t) {
+      logError("Exceptions must not be thrown from onError", t);
+    }
+  }
+
+  private void sendComplete() {
+    try {
+      cancelled = true;
+      subscriber.onComplete();
+    } catch (Throwable t) {
+      logError("Exceptions must not be thrown from onComplete", t);
+    }
+  }
+
+  private void doSubscribe(final Subscriber<? super T> subscriber) {
+    this.subscriber = subscriber;
+    try {
+      subscriber.onSubscribe(new Sub());
+    } catch (final Throwable t) {
+      sendError(new IllegalStateException("Exceptions must not be thrown from onSubscribe", t));
+    }
+  }
+
+  private void checkContext() {
+    if (Vertx.currentContext() != ctx) {
+      throw new IllegalStateException("On wrong context");
+    }
+  }
+
+  private void doSend() {
+    int numSent = 0;
+    while (!cancelled && demand > 0 && !buffer.isEmpty()) {
+      if (numSent < SEND_MAX_BATCH_SIZE) {
+        final T val = buffer.poll();
+        doOnNext(val);
+        numSent++;
+      } else {
+        // Schedule another batch async
+        ctx.runOnContext(v -> doSend());
+        break;
+      }
+    }
+
+    if (buffer.isEmpty() && !cancelled) {
+      if (complete) {
```

Comment:
is it possible that we will never send the complete? it seems like if subscriber that was set never calls `request`, we won't increase our demand and we will never the buffered elements

Reply:
This is by design, the complete will only be sent after all buffered messages have been delivered, and they won't be delivered unless they've been requested.